File: 36-tele-book/tele_book.py
# Telephone Directory book program in Python
# Written by M.V.Harish Kumar on 18-06-2025
import json
import sys

telebook = {}

def print_banner(data, title=False):
    print(data)
    print('*' * len(data) if title else '-' * len(data))

def add():
    print_banner("Add contact")
    name = input("Enter contact name: ")
    telebook[name] = input("Enter contact no: ")
    print("Added contact successfully")

def view():
    print_banner("View contact")
    if not telebook:
        print("Contact list is empty!")
        return
    for k in telebook:
        print(f"- {k}")
    cname = input("Select contact: ")
    num = telebook.get(cname, "Not found")
    print(f"Contact details for {cname}: {num}")

def update():
    print_banner("Update contact")
    if not telebook:
        print("Contact list is empty!")
        return
    for k, v in telebook.items():
        print(f"- {k}: {v}")
    cname = input("Select contact to update: ")
    try:
        telebook[cname] = input(f"Enter new number for {telebook[cname]}: ")
        print("Updated contact successfully")
    except KeyError:
        print("Attempt to access non-existant contact")

def delete():
    print_banner("Delete contact")
    if not telebook:
        print("Contact list is empty!")
        return
    for k, v in telebook.items():
        print(f"- {k}: {v}")
    cname = input("Select contact to delete(or * for all): ")
    if cname == '*':
        telebook.clear()
        print("Cleared all data")
    else:
        try:
            del telebook[cname]
            print("Deleted data successfully")
        except KeyError:
            print("Attempt to access non-existant contact")

def persist():
    if not telebook:
        print("Contact list is empty! No contact to save")
        return
    print_banner("Persist data")
    fname = input("Give the file name to save: ")
    with open(fname, 'w') as f:
        json.dump(telebook, f)


if __name__ == '__main__':
    if len(sys.argv) == 2:
        with open(sys.argv[1]) as f:
            telebook = json.load(f)
    elif len(sys.argv) != 1:
        raise TypeError("Invalid no of args")

    tele_switch = [add, view, update, delete, persist, exit]

    print_banner("Welcome to TeleBook", title=True)
    while True:
        print("1. Add contact")
        print("2. View contact")
        print("3. Update contact")
        print("4. Delete contact(s)")
        print("5. Persist telebook")
        print("6. Exit")

        opt = int(input("Enter an option: "))
        try:
            tele_switch[opt-1]()
        except IndexError:
            print("Invalid option")

File: 36-tele-book/tele_book.go
// Telephone directory book program in Go
// Written by M.V.Harish Kumar on 19-06-2025
package main

import (
	"bufio"
	"encoding/json"
	"fmt"
	"os"
	"regexp"
	"strings"
)

var telebook map[string]string
var scnr = bufio.NewScanner(os.Stdin)

func isValidNumber(number string) bool {
	re := regexp.MustCompile(`^(\+?91)?\d{10}`)
	return re.MatchString(number)
}

func printBanner(msg string, isTitle bool) {
	fmt.Println(msg)
	char := "-"
	if isTitle {
		char = "*"
	}
	fmt.Println(strings.Repeat(char, len(msg)))
}

func inputString(prompt string) string {
	fmt.Print(prompt)
	scnr.Scan()
	return scnr.Text()
}

func add() {
	printBanner("Add Contact", false)
	name := inputString("Enter contact name: ")
	number := inputString("Enter contact number: ")
	if isValidNumber(number) {
		telebook[name] = number
		fmt.Println("Added contact successfully")
	} else {
		fmt.Println("Invalid number format")
	}
}

func view() {
	printBanner("View contacts", false)
	if len(telebook) == 0 {
		fmt.Println("Contact list is empty!")
		return
	}
	for k := range telebook {
		fmt.Printf("- %s\n", k)
	}
	cname := inputString("Select contact: ")
	num, hasName := telebook[cname]
	if !hasName {
		num = "Not found"
	}
	fmt.Printf("Contact details for %s: %s\n", cname, num)
}

func update() {
	printBanner("Update contacts", false)
	if len(telebook) == 0 {
		fmt.Println("Contact list is empty!")
		return
	}
	for k, v := range telebook {
		fmt.Printf("- %s: %s\n", k, v)
	}
	cname := inputString("Select contact to update: ")
	_, hasName := telebook[cname]
	if !hasName {
		fmt.Println("Attempt to access non-existant contact")
		return
	}
	newNum := inputString(fmt.Sprintf("Enter new number for %s: ", cname))
	if isValidNumber(newNum) {
		telebook[cname] = newNum
		fmt.Println("Updated contact successfully")
	} else {
		fmt.Println("Invalid number format")
	}
}

func del() {
	printBanner("Delete contact", false)
	if len(telebook) == 0 {
		fmt.Println("Contact list is empty!")
		return
	}
	for k, v := range telebook {
		fmt.Printf("- %s: %s\n", k, v)
	}
	cname := inputString("Select contact to delete(or * for all): ")
	if cname == "*" {
		for k := range telebook {
			delete(telebook, k)
		}
		fmt.Println("Cleared all data")
	} else {
		if _, ok := telebook[cname]; !ok {
			fmt.Println("Attempt to delete non-existant contact")
			return
		}
		delete(telebook, cname)
		fmt.Println("Deleted contact successfully")
	}
}

func persist() {
	if len(telebook) == 0 {
		fmt.Println("Contact list is empty! No contact to save")
		return
	}
	printBanner("Persist data", false)
	fname := inputString("Enter file name to save: ")
	data, _ := json.Marshal(telebook)
	os.WriteFile(fname, data, 0666)
	fmt.Println("Saved to file successfully")
}

func main() {
	var opt uint
	if len(os.Args) == 2 {
		ftxt, _ := os.ReadFile(os.Args[1])
		json.Unmarshal(ftxt, &telebook)
	} else if len(os.Args) == 1 {
		telebook = make(map[string]string)
	} else {
		panic("Invalid no of arguments")
	}

	teleswitch := [5]func(){add, view, update, del, persist}

	printBanner("Welcome to telebook", true)
	for {
		fmt.Println("1. Add contact")
		fmt.Println("2. View contact")
		fmt.Println("3. Update contact")
		fmt.Println("4. Delete contact(s)")
		fmt.Println("5. Persist telebook")
		fmt.Println("6. Exit")

		fmt.Print("Enter an option: ")
		fmt.Scanln(&opt)

		if opt == 6 {
			break
		} else if opt > 0 && opt < 6 {
			teleswitch[opt-1]()
		} else {
			fmt.Println("Invalid option")
		}
	}
}
